openhands/controller/agent_controller.py

```diff
@@ -105,18 +105,19 @@ def __init__(
         self.agent = agent
         self.headless_mode = headless_mode
 
-        # subscribe to the event stream
-        self.event_stream = event_stream
-        self.event_stream.subscribe(
-            EventStreamSubscriber.AGENT_CONTROLLER, self.on_event, self.id
-        )
-
         # state from the previous session, state from a parent agent, or a fresh state
         self.set_initial_state(
             state=initial_state,
             max_iterations=max_iterations,
             confirmation_mode=confirmation_mode,
         )
+
+        # subscribe to the event stream
+        self.event_stream = event_stream
+        self.event_stream.subscribe(
+            EventStreamSubscriber.AGENT_CONTROLLER, self.on_event, self.id
+        )
+
         self.max_budget_per_task = max_budget_per_task
```

Comment:
Why change the order here?

Reply:
I suppose it's probably not necessary - One of the main issues I was getting as part of this PR was that the agent was repeating the actions on session resume. I wanted to make sure that the event stream subscribers are only added (And can only be alerted) after the state is restored.

Reply:
I mean this would fail: https://github.com/All-Hands-AI/OpenHands/blob/27716171bf56edf2f539b5da4becf0cec066937b/openhands/controller/agent_controller.py#L674